include/nbl/ui/IInputEventChannel.h

```diff
@@ -0,0 +1,108 @@
+#ifndef __NBL_I_INPUT_EVENT_CHANNEL_H_INCLUDED__
+#define __NBL_I_INPUT_EVENT_CHANNEL_H_INCLUDED__
+
+#include <mutex>
+
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/core/containers/CCircularBuffer.h"
+#include "nbl/core/SRange.h"
+
+namespace nbl {
+namespace ui
+{
+
+class IInputEventChannel : public core::IReferenceCounted
+{
+protected:
+    virtual ~IInputEventChannel() = default;
+
+public:
+    // TODO Any other common (not depending on event struct type) member functions?
+
+    virtual bool empty() const = 0;
+};
+
+namespace impl
+{
+    template <typename EventType>
+    class IEventChannelBase : public IInputEventChannel
+    {
+    protected:
+        static inline constexpr size_t MaxEvents = 1024u;
+
+        using cb_t = core::CCompileTimeSizedCircularBuffer<EventType, MaxEvents>;
+        using iterator_t = typename cb_t::iterator;
+        using range_t = core::SRange<EventType, iterator_t, iterator_t>;
+
+        std::mutex m_bgEventBufMtx;
+        cb_t m_bgEventBuf;
+        cb_t m_frontEventBuf;
+
+        // Lock while working with background event buffer
+        std::unique_lock<std::mutex> lockBackgroundBuffer()
+        {
+            return std::unique_lock<std::mutex>(m_bgEventBufMtx);
+        }
+
+        // Use this within OS-specific impl (Windows callback/XNextEvent loop thread/etc...)
+        void pushIntoBackground(EventType&& ev)
+        {
+            m_bgEventBuf.push_back(std::move(ev));
+        }
```

Comment:
for "pumping" virtual events (ones that we make up, or forward from middlewares/other engines), we need a `mergeIntoForeground(begin,end)` where we do a simple mergesort step, mergining the list of events already present in the foreground cbuffer and the list of events given by the arguments `begin,end`

We can assume both lists/ranges are sorted, hence we just need to merge them (while handling overflow properly).

Reply:
there's probably an `std::merge` or something for this